File: frontend/be_elite/lib/models/Session/Athlete Sessions/athlete_session_dto.dart
import 'package:be_elite/models/Session/Athlete%20Sessions/athlete_session_id.dart';
import 'package:be_elite/models/Session/Athlete%20Sessions/block.dart';

class AthleteSessionDto {
  AthleteSessionId? sessionId;
  String? date;
  String? title;
  String? subtitle;
  int? sameDaySessionNumber;
  List<AthleteBlockDto>? blocks;
  bool? isCompleted;

  AthleteSessionDto({
    this.sessionId,
    this.date,
    this.title,
    this.subtitle,
    this.sameDaySessionNumber,
    this.blocks,
    this.isCompleted,
  });

  factory AthleteSessionDto.fromJson(Map<String, dynamic> json) {
    return AthleteSessionDto(
      sessionId: json['sessionId'] == null
          ? null
          : AthleteSessionId.fromJson(json['sessionId'] as Map<String, dynamic>),
      date: json['date'] as String?,
      title: json['title'] as String?,
      subtitle: json['subtitle'] as String?,
      sameDaySessionNumber: json['sameDaySessionNumber'] as int?,
      blocks: (json['blocks'] as List<dynamic>?)
          ?.map((e) => AthleteBlockDto.fromJson(e as Map<String, dynamic>))
          .toList(),
      isCompleted: json['isCompleted'] as bool?,
    );
  }

  Map<String, dynamic> toJson() => {
        'sessionId': sessionId?.toJson(),
        'date': date,
        'title': title,
        'subtitle': subtitle,
        'sameDaySessionNumber': sameDaySessionNumber,
        'blocks': blocks?.map((e) => e.toJson()).toList(),
        'isCompleted': isCompleted,
      };
}

File: backend/BeElite/src/main/java/com/salesianos/triana/BeElite/dto/Session/AthleteSessionDto.java
package com.salesianos.triana.BeElite.dto.Session;

import com.salesianos.triana.BeElite.dto.Block.AthleteBlockDto;
import com.salesianos.triana.BeElite.model.AthleteSession;
import com.salesianos.triana.BeElite.model.Composite_Ids.AthleteSessionId;
import com.salesianos.triana.BeElite.model.Composite_Ids.SessionId;
import lombok.Builder;
import lombok.Getter;

import java.time.LocalDate;
import java.util.List;

@Builder
public record AthleteSessionDto(
        AthleteSessionId sessionId,
        LocalDate date,
        String title,
        String subtitle,
        int sameDaySessionNumber,
        List<AthleteBlockDto> blocks,
        boolean isCompleted
) {
    public static AthleteSessionDto of(AthleteSession a){
        return AthleteSessionDto.builder()
                .sessionId(AthleteSessionId.of(a.getId().getAthlete_session_number(), a.getId().getAthlete_id()))
                .date(a.getSession().getDate())
                .title(a.getSession().getTitle())
                .subtitle(a.getSession().getTitle())
                .sameDaySessionNumber(a.getSession().getSame_day_session_number())
                .blocks(a.getAthleteBlocks() == null ? List.of() : a.getAthleteBlocks().stream().map(AthleteBlockDto::of).toList())
                .isCompleted(a.isCompleted())
                .build();
    }

}
